File: packages/javascript/lib/node/net/fetchData.ts
import type { AnyValueObject, HTTPRequestOption } from '../../_types/global';
import { objToQueryString } from '../../object/objToQueryString.js';
import { urlJoin } from '../../string/urlJoin.js';

function generateDefaultRequestHeader(httpRequestOption?: HTTPRequestOption) {
	let contentType;

	if (httpRequestOption?.headers?.['Content-Type']) {
		contentType = httpRequestOption?.headers?.['Content-Type'];
	} else {
		switch (httpRequestOption?.bodyType) {
			case 'form-data':
				contentType = 'multipart/form-data';
				break;
			case 'x-www-form-urlencoded':
				contentType = 'application/x-www-form-urlencoded';
				break;
			case 'json':
				contentType = 'application/json;charset=UTF-8';
				break;
			case 'text':
				contentType = 'plain/text;charset=UTF-8';
				break;
			default:
				contentType = null;
				break;
		}
	}

	const authHeaders: AnyValueObject = {};

	if (httpRequestOption?.auth?.bearer && httpRequestOption?.auth?.bearer.length > 0) {
		authHeaders.Authorization = `Bearer ${httpRequestOption?.auth?.bearer}`;
	}
	if (httpRequestOption?.auth?.apiKey && httpRequestOption?.auth?.apiKey.length > 0) {
		authHeaders['x-API-key'] = httpRequestOption?.auth?.apiKey;
	}

	return {
		...authHeaders,
		...httpRequestOption?.headers,
		...(contentType ? { 'Content-Type': contentType } : {}),
		'Accept-Encoding': 'gzip, deflate, br',
		Charset: 'utf-8'
	};
}

export async function fetchData(url: string, options?: HTTPRequestOption): Promise<any> {
	const opt = options || {};
	const queryString = opt.queryParameters ? `?${objToQueryString(opt.queryParameters)}` : null;
	let bodyData = null;

	if (url.length < 1) {
		throw new Error('`url` is required');
	}

	if (opt.body) {
		switch (opt.bodyType) {
			case 'x-www-form-urlencoded':
				bodyData = objToQueryString(opt.body as object);
				break;
			case 'form-data':
				bodyData = opt.body as FormData;
				break;
			case 'json':
			default:
				bodyData = JSON.stringify(opt.body as object);
				break;
		}
	}

	let method;

	if (opt.method && (opt.get || opt.post || opt.put || opt.delete || opt.patch)) {
		throw new Error('`method` and `get|post|put|delete|patch` cannot be used together');
	}

	if (opt.get) {
		method = 'GET';
	} else if (opt.post) {
		method = 'POST';
	} else if (opt.put) {
		method = 'PUT';
	} else if (opt.delete) {
		method = 'DELETE';
	} else if (opt.patch) {
		method = 'PATCH';
	} else {
		method = opt.method || 'GET';
	}

	let fullRequestUrl;

	if (!opt.host && !url.startsWith('/')) {
		throw new Error('`url` must begin with `/`.');
	}

	if (opt.host && opt.host.length > 0) {
		if (url.startsWith('http') || url.includes('://')) {
			throw new Error('If `host` is specified, `url` must begin with `/`.');
		}

		fullRequestUrl = urlJoin(opt.host, url);
	} else {
		fullRequestUrl = url;
	}

	try {
		const response = await fetch(urlJoin(fullRequestUrl, queryString), {
			...(opt.auth ? { credentials: 'include' } : {}),
			method,
			body: bodyData,
			headers: opt.headers || generateDefaultRequestHeader(opt)
		});

		if (!response.ok || !response.body) {
			return null;
		}

		const responseContentType = response.headers.get('content-type');

		if (!responseContentType) {
			return await response.text();
		} else if (responseContentType.includes('application/json')) {
			return await response.json();
		} else {
			return await response.text();
		}
	} catch (error: any) {
		if (opt.onError) {
			opt.onError(error);
		}

		return null;
	}
}

File: packages/python/qsu/net/fetchData.py
import json
import urllib.request
import urllib.error

from ..string.urlJoin import urlJoin
from ..object.objToQueryString import objToQueryString


def _generateDefaultRequestHeader(opt):
	contentType = None

	headers = opt.get('headers') or {}

	if headers.get('Content-Type'):
		contentType = headers.get('Content-Type')
	else:
		bodyType = opt.get('bodyType')

		if bodyType == 'form-data':
			contentType = 'multipart/form-data'
		elif bodyType == 'x-www-form-urlencoded':
			contentType = 'application/x-www-form-urlencoded'
		elif bodyType == 'json':
			contentType = 'application/json;charset=UTF-8'
		elif bodyType == 'text':
			contentType = 'plain/text;charset=UTF-8'
		else:
			contentType = None

	result = {}

	auth = opt.get('auth') or {}

	if auth.get('bearer') and len(auth.get('bearer')) > 0:
		result['Authorization'] = f"Bearer {auth.get('bearer')}"
	if auth.get('apiKey') and len(auth.get('apiKey')) > 0:
		result['x-API-key'] = auth.get('apiKey')

	if contentType:
		result['Content-Type'] = contentType

	result['Accept-Encoding'] = 'gzip, deflate, br'
	result['Charset'] = 'utf-8'

	for key in headers.keys():
		result[key] = headers[key]

	return result


def _isFileResponse(contentType: str, contentDisposition: str) -> bool:
	import re

	if re.search(r'attachment|filename=', contentDisposition, re.IGNORECASE):
		return True

	mime = (contentType.split(';')[0].strip().lower()) if contentType else ''

	textTypes = [
		'text/html',
		'text/css',
		'text/javascript',
		'application/json',
		'application/ld+json',
		'application/xml',
		'text/xml',
		'text/plain',
	]

	if any(mime == t for t in textTypes) or mime.startswith('text/'):
		return False

	return (
		mime.startswith('application/')
		or mime.startswith('image/')
		or mime.startswith('video/')
		or mime.startswith('audio/')
		or mime.startswith('font/')
	)


def fetchData(url: str, options=None):
	opt = options or {}
	queryString = (
		f'?{objToQueryString(opt.get("queryParameters"))}'
		if opt.get('queryParameters')
		else None
	)
	bodyData = None

	if len(url) < 1:
		raise ValueError('`url` is required')

	if opt.get('body'):
		bodyType = opt.get('bodyType')

		if bodyType == 'x-www-form-urlencoded':
			bodyData = objToQueryString(opt.get('body'))
		elif bodyType == 'form-data':
			bodyData = opt.get('body')
		else:
			bodyData = json.dumps(opt.get('body'))

	if opt.get('method') and (
		opt.get('get')
		or opt.get('post')
		or opt.get('put')
		or opt.get('delete')
		or opt.get('patch')
	):
		raise ValueError(
			'`method` and `get|post|put|delete|patch` cannot be used together'
		)

	if opt.get('get'):
		method = 'GET'
	elif opt.get('post'):
		method = 'POST'
	elif opt.get('put'):
		method = 'PUT'
	elif opt.get('delete'):
		method = 'DELETE'
	elif opt.get('patch'):
		method = 'PATCH'
	else:
		method = (opt.get('method') or 'GET').upper()

	if not opt.get('host') and not url.startswith('/'):
		raise ValueError('`url` must begin with `/`.')

	if opt.get('host') and len(opt.get('host')) > 0:
		if url.startswith('http') or '://' in url:
			raise ValueError('If `host` is specified, `url` must begin with `/`.')

		fullRequestUrl = urlJoin(opt.get('host'), url)
	else:
		fullRequestUrl = url

	try:
		requestUrl = urlJoin(fullRequestUrl, queryString)

		data = None

		if bodyData is not None:
			data = (
				bodyData.encode('utf-8')
				if isinstance(bodyData, str)
				else bodyData
			)

		request = urllib.request.Request(
			requestUrl,
			data=data,
			method=method,
			headers=_generateDefaultRequestHeader(opt),
		)

		timeout = opt.get('timeout')
		timeoutSeconds = (timeout / 1000) if timeout else None

		with urllib.request.urlopen(request, timeout=timeoutSeconds) as response:
			status = response.status
			rawBody = response.read()

			if status < 200 or status >= 300 or status == 204 or not rawBody:
				return None

			responseContentType = response.headers.get('content-type') or ''
			contentDisposition = response.headers.get('content-disposition') or ''
			isFile = _isFileResponse(responseContentType, contentDisposition)

			if opt.get('toStream'):
				return rawBody
			elif isFile:
				return rawBody
			elif len(responseContentType) < 1:
				return rawBody.decode('utf-8')
			elif 'application/json' in responseContentType:
				return json.loads(rawBody.decode('utf-8'))
			else:
				return rawBody.decode('utf-8')
	except Exception as error:
		if opt.get('onError'):
			opt.get('onError')(error)

		return None
